### ISSUE TITLE
React to ScrollWheel in from-event

### ISSUE THREAD

**Issue by wedi** [tekezo/Karabiner-Elements#1362]
Hi,
I got into [the habbit](https://xkcd.com/1172/) using <kbd>ctrl</kbd> + <kbd>left/right scroll</kbd> to switch desktops. Unfortunately, I still haven't found a way to achieve this with Karabiner Elements.

Does anyone know a way to make this work? I am looking for a way to react to horizontal ScrollWheel events. [As *to-event* this is documented](https://pqrs.org/osx/karabiner/json.html#to-event-definition) but unfortunately I haven't found anything about a correspondig *from-event*.

Thank you!

Related Issues:
https://github.com/tekezo/Karabiner-Elements/issues/261#issuecomment-339164627, #579, #1066, #1113, https://github.com/tekezo/Karabiner/issues/814

In the old days it used to work this way:
```xml
<autogen>
    __ScrollWheelToKey__
    ScrollWheel::RIGHT, MODIFIERFLAG_EITHER_LEFT_OR_RIGHT_CONTROL,
    KeyCode::CURSOR_LEFT,  MODIFIERFLAG_EITHER_LEFT_OR_RIGHT_CONTROL,
</autogen>
```

#1066 describes exactly how this was achieved earlier:
> The previous version of Karabiner allowed one to reference ScrollWheel::Left and ScrollWheel::Right in the "from" field of mappings. I have not been able to find any documentation on how to do this in Karabiner-Elements.

> In Karabiner-EventViewer I just see a "ScrollWheel" event come in along with "misc" data: "dx: -2.10 dy: 0.00 dz: 0:00".

**Comment by nathanko711** [comment]
Hey I would love to know this as well, using left/right scroll to map to back/forward buttons would be pretty awesome.

**Comment by wedi** [comment]
Just stumbled upon one more related issue: #1314
There is a simple modification TO a mouse scroll event but not FROM:

![from_mouse_key](https://user-images.githubusercontent.com/1466269/40670144-f1a5f9f4-6368-11e8-8a73-84dac584600f.png)
_From key_

![to_mouse_key](https://user-images.githubusercontent.com/1466269/40670149-f50f6cf6-6368-11e8-92e9-47a5f3dde00b.png)
_To key_

Also see this (closed) pull request: #1230

**Comment by wedi** [comment]
Is there absolutely no solution to remap left/right scrolling? Not even a slight hint? What a pity.

**Comment by hakanersu** [comment]
Yes i found that in old versions there is a way to do that which i really need that with my sculpt mouse.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root>
  <devicevendordef>
    <vendorname>MICROSOFT</vendorname>
    <vendorid>0x045e</vendorid>
  </devicevendordef>
  <deviceproductdef>
    <productname>MICROSOFT_SCULPT_MOUSE</productname>
    <productid>0x07a2</productid>
  </deviceproductdef>
  <item>
    <name>Microsoft Sculpt Mouse - Map Windows Button</name>
    <identifier>private.msft_sculpt</identifier>
    <device_only>
        DeviceVendor::MICROSOFT, 
        DeviceProduct::MICROSOFT_SCULPT_MOUSE
    </device_only>
    <autogen>
        __KeyToKey__
        KeyCode::CONTROL_L,
        KeyCode::VK_NONE,
        ModifierFlag::NONE
    </autogen>
    <autogen>
        __KeyOverlaidModifier__
        KeyCode::COMMAND_L,
        KeyCode::COMMAND_L,
        KeyCode::LAUNCHPAD
    </autogen>
    <autogen>
        __KeyToKey__
        KeyCode::DELETE,
        ModifierFlag::COMMAND_L,
        KeyCode::MISSION_CONTROL
    </autogen>
    <autogen>
        __KeyToKey__
        KeyCode::TAB,
        ModifierFlag::COMMAND_L,
        KeyCode::CURSOR_DOWN,
        ModifierFlag::CONTROL_L
    </autogen>
    <autogen>
        __ScrollWheelToKey__
        ScrollWheel::LEFT,
        ModifierFlag::COMMAND_L,
        KeyCode::CURSOR_LEFT,
        ModifierFlag::CONTROL_L
    </autogen>
    <autogen>
        __ScrollWheelToKey__
        ScrollWheel::RIGHT,
        ModifierFlag::COMMAND_L,
        KeyCode::CURSOR_RIGHT,
        ModifierFlag::CONTROL_L
    </autogen>
    <autogen>
        __FlipScrollWheel__
        Option::FLIPSCROLLWHEEL_VERTICAL,
        Option::FLIPSCROLLWHEEL_HORIZONTAL
    </autogen>
  </item>
</root>
```

**Comment by wedi** [comment]
Thanks for the hint! Unfortunately this is a configuration for the legacy Karabiner (as one can tell from the XML format) not the current Karabiner Elements. The new version has no *obvious* support for this and I couldn't find a workaround yet.

**Comment by karlapsite** [comment]
Bump: wouldn't mind being able to rebind scroll wheel to perform certain actions.

I'm trying to zoom in and out using a scroll wheel in Zeplin, so I need to rebind wheel up to ⌘+ and wheel down to ⌘-.

**Comment by calvinwyoung** [comment]
Has anyone found a workaround for this? I'm also looking for a solution right now.

**Comment by zgeist27** [comment]
I'm looking for this, too.

**Comment by FuegoFro** [comment]
To feed the stale bot, I'll chime in that I'd also love for this feature to be implemented.

**Comment by chavlji** [comment]
I also desperately need wheel up/down as from key. I need it to:
- switch tabs with right mouse button down + wheel
- switch apps with mouse button 4 down + wheel
I'm begging you please!

**Comment by sterling** [comment]
Still a pretty significant missing piece of functionality.

**Comment by krzysztofwolski** [comment]
As a new MS Sculpt user I'm also interested in this func.

**Comment by RainbowTraveller** [comment]
Yes. Need ability to scroll using scroll wheel....!

**Comment by piotr-tobolski** [comment]
Any news?

**Comment by stianhoiland** [comment]
I would like to see this, too!

**Comment by yanickxia** [comment]
waiting for it....

**Comment by kpau** [comment]
+1

**Comment by fogrid** [comment]
Yea, I would like this feature too - I really want to be able to zoom with the mouse

**Comment by prasskaly** [comment]
I would like to have the feature of Ctrl+scroll up/down to zoom in/zoom out. Firefox supports this mode for macOS, but I want it to natively run in all apps that will be only possible with a tool like Karabiner.

**Comment by kerryj89** [comment]
Since it's already understood in the to-event it should be trivial to understand from the from-event, no?  Of course you and I both know things aren't always as simple as they seem.

**Comment by the-eric-kwok** [comment]
+1

**Comment by huseyinyagliudemy** [comment]
I'm missing this feature to reverse my mouse scroll without reversing trackpad scroll.

**Comment by donmurdoc** [comment]
I'm going crazy trying to do this!

**Comment by catkic** [comment]
+1

**Comment by Telemakhos** [comment]
+1

**Comment by stephancasas** [comment]
+1

**Comment by mam17** [comment]
+1

**Comment by cipri-tom** [comment]
for people needing versatile Mouse events, I would recommend USB Overdrive. I'm new to mac (2 years) and started using that before discovering Karabiner, as I needed to keep natural scroll on the trackpad and not on the mouse. Now I'm using both: Karabiner for keyboard, USB Overdrive for mouse.

While it is not free, it is nag-ware. You have to wait 30 seconds when you want to enter the settings panel. I've used it like that for 2 years without a problem (I paid it now that I have a decent salary).

**Comment by ibehnam** [comment]
Still waiting for this functionality in 2020...

**Comment by wedi** [comment]
This ticket is getting close to the top in the most commented list. 😆

@tekezo Could you give an estimate if you are planning to implement this at some time or if we should just use an alternative like USB Overdrive for this particular feature?

Please get me right, this is an open source project you do in your spare time and I am really grateful for what I get right now. It's just for us/me to know.

**Comment by drhisham-code** [comment]
really miss this functionality

**Comment by drhisham-code** [comment]
bump

**Comment by drhisham-code** [comment]
bump

**Comment by dts88** [comment]
Yes. I do need to have a react to Scroll_Wheel in from-event too. 
Now it just has "event type" with misc. No "name" or "code". can not control it.

**Comment by cjh9** [comment]
This would be awesome if this functionality was implemented :)

**Comment by radman0x** [comment]
I've found my way here too, want to be able to map the mouse wheel + a modifier to zoom apps in and out. Hopefully there's time to look at this feature at some point :)

**Comment by cipri-tom** [comment]
@radman0x I understand your need :D. Most apps can already do that (web brosers, code editors, etc). This is great because they can zoom in "smartly" depending on their content.

Otherwise, on mac, you can enable an accessibility feature that zoom the whole screen: Settings > Accessibility > Zoom > _Use scroll gesture with modifier keys to zoom_ . Enable this, and select the modifier you want (Ctrl, Option, Cmd). There are also 2 styles of zoom, see which one you like better.

**Comment by Trinistar** [comment]
+1

**Comment by epjnano** [comment]
+1, trying to find any method at all to remap scrolls on Mac, and Karabiner is the closest thing I've found (USB Overdrive doesn't cut it since I'm not using a USB/BT mouse)...would really love to see this implemented!

**Comment by uniqueidentifier** [comment]
I see scrollwheel functoinality is on the road map.

Maybe this will be useful for someone. I'm uploading Swift code I hobbled together today to make a command line program I can integrate with a Custom Command

[main.swift.txt](https://github.com/pqrs-org/Karabiner-Elements/files/5790672/main.swift.txt)

What the program does is [1] listen for scroll events, and convert them to left and right arrow keypresses (a more sophisticated version would actually throttle according to the deltaY values, but I hacked this together in a  couple hours, so I didn't bother) [2] as soon as the user creates any keypress or click, the program exits.

My Complex Karabiner command to start it up looks like this (when the user presses F16, Karabiner presses Fn F2 to select the Apple Menu, and then runs my helper tool so the user can scroll left or right through the Menu bar)

		{ 
			"description": "F16 ➞ Menu", "manipulators": [
				{
				"type": "basic",
				"from": { "key_code": "f16" },
				"to": [
						{ "key_code": "f2", "modifiers": ["control"] },
						{ "key_code": "down_arrow", "repeat": false }
					],
				"to_after_key_up": [
						{ "shell_command": "~/bin/scroll_to_left_right_arrows" }
					]
				}
			]
		}

**Comment by DKroot** [comment]
Please, please... Accidental zooms with Crtl+Scroll on Mac drive me nuts. Even though this "feature" can be turned off in Accessibility, apps like Chrome and Excel still do it with no easy way to disable (there is a Chrome extension for it). I'd love to be able to disable this system-wide.

**Comment by wedi** [comment]
As @uniqueidentifier posted, @tekezo [has put it](https://github.com/pqrs-org/Karabiner-Elements/commit/69126c945fcc54116722694b31acdbf9241ef5f9) on the [roadmap](https://github.com/pqrs-org/Karabiner-Elements/blob/master/TASKS.md) but I wouldnt' be too optimistic about it speeding up this issue. 😉 Personally, I've made my peace with this situation as Karabiner is great without this feature already but I totally understand people not being happy about this.

**Comment by Loakesy** [comment]
+99

**Comment by scrapix** [comment]
Mouse Wheel Functions would be great indeed, maybe we can support with some donations :)

**Comment by WDmoose** [comment]
+1 bump bump, would love to see this added.

**Comment by nonsonodavid** [comment]
i have the same problem, but the point is understand what type of input is the "scroll_wheel", a keycode? a keyconsumercode? or a pointing buttom? Sorry for my english, is bad, but this can help everyone to solve the problem.

**Comment by 03k64serenity** [comment]
This would be an excellent addition.

**Comment by codeTIT4N** [comment]
eagerly waiting for this feature

**Comment by wedi** [comment]
It might help if more people upvoting this feature request would sponsor @tekezo ... ;)

**Comment by kglueck** [comment]
+1

**Comment by thetimetraveller** [comment]
> I noticed that 'scrollwheel functionality' is slated on the road map.
> 
> Maybe the following will be useful for someone: attached is Swift code I hobbled together today to make a tiny little 'scroll_to_left_right_arrows' CLI program. A user can launch such a program as part of a Karabiner Custom Command. I think this workaround can implement many of the use-cases for the requested feature (until such time as dev implements the feature).
> 
> [main.swift.txt](https://github.com/pqrs-org/Karabiner-Elements/files/5790672/main.swift.txt)
> 
> What my "scroll_to_left_right_arrows" program does is [1] listen for scroll events, and convert them to left and right arrow keypresses (a more sophisticated version would actually throttle according to the deltaY values, but I hacked this together in a couple hours, so I didn't bother) [2] as soon as the user creates any keypress or click, the program exits (which, if anyone edits my program,they should be mindful of, since without exiting, you won't be able to operate any thing at all except left and right arrow! At one point today I absent-mindedly made a version that didn't exit, and had to reboot my Mac!).
> 
> My Complex Karabiner command to start it up looks like this (when the user presses F16, Karabiner presses Fn F2 to select the Apple Menu, and then runs my helper tool so the user can scroll left or right through the Menu bar)
> 
> ```
> 	{ 
> 		"description": "F16 ➞ Menu", "manipulators": [
> 			{
> 			"type": "basic",
> 			"from": { "key_code": "f16" },
> 			"to": [
> 					{ "key_code": "f2", "modifiers": ["control"] },
> 					{ "key_code": "down_arrow", "repeat": false }
> 				],
> 			"to_after_key_up": [
> 					{ "shell_command": "~/bin/scroll_to_left_right_arrows" }
> 				]
> 			}
> 		]
> 	} 
> ```

This sounds very good!

How can I use it inside a complex modification?

I can't run your script in terminal to test it: 'import: command not found' among other errors :(

**Comment by MuhammedZakir** [comment]
> I can't run your script in terminal to test it: 'import: command not found' among other errors :(

It's a Swift code, not shell script. So, you will have to compile it first. After that, use the executable you got.

> How can I use it inside a complex modification?

There is an example in the post you quoted. `~/bin/scroll_to_left_right_arrows` is the executable. Your path may differ depending on where you compile it, or put the executable.

**Comment by thetimetraveller** [comment]
> > I can't run your script in terminal to test it: 'import: command not found' among other errors :(
> 
> It's a Swift code, not shell script. So, you will have to compile it first. After that, use the executable you got.
> 
> > How can I use it inside a complex modification?
> 
> There is an example in the post you quoted. `~/bin/scroll_to_left_right_arrows` is the executable. Your path may differ depending on where you compile it, or put the executable.

Hi Muhammed, thanks for the answer.

I tried to compile the script in Terminal (version of Swift 5.1.3, but I get this error:


`main.swift:118:84: error: the compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions
                        = (1 << 1) | (1 << 3) | (1 << 6) | (1 << 7) | (1 << 10) | (1 << 22) | (1 << 25) | (1 << 27)
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
`

Maybe is there a typo in the script?


Best!

**Comment by MuhammedZakir** [comment]
Try splitting that line:
```swift
// Split the expression due to compilation error.
let _mask1: UInt32 = (1 << 1) | (1 << 3) | (1 << 6)
let _mask2: UInt32 = _mask1 | (1 << 7) | (1 << 10)
let mask: UInt32 = _mask2 | (1 << 22) | (1 << 25) | (1 << 27)
```

**Comment by plesui** [comment]
🙏

**Comment by rvanmazijk** [comment]
+1

**Comment by deadElk** [comment]
need both scroll wheels func (mx master 3) 
ready to donate

**Comment by wedi** [comment]
> need both scroll wheels func (mx master 3)
> ready to donate
 – @deadElk 

There is this old saying: _Actions speak louder than words._ 

![grafik](https://user-images.githubusercontent.com/1466269/144607049-59d6c760-8a42-415c-9cf0-985ef4d9e584.png)

could show

![grafik](https://user-images.githubusercontent.com/1466269/144606862-6e334cec-8933-4171-81df-249a9dae990e.png)

Maybe that helps supporting this issue. I'd be very happy. 😃

**Comment by alexDevBR** [comment]
@radman0x and anyone wanting to zoom with ctrl+scroll I modified the swift code provided by @uniqueidentifier to do that.
[main.swift.txt](https://github.com/pqrs-org/Karabiner-Elements/files/9074493/main.swift.txt)
Compile it with: `swiftc main.swift -o scrollzoom`
I don't know swift so it's a hack based on a hack, use at your own risk!

**Comment by BalazsGyarmati** [comment]
Based on @alexDevBR and @uniqueidentifier 's code I've added keyboard layout independency (original code worked only with US layout), changed the control key to command key and fixed a bug (the scroll delta value for my mouse was 0 in idle state, and the delta_y > 0 wasn't enough for deciding on zoom direction)

Uploaded the fresh code here:
https://gist.github.com/BalazsGyarmati/e199080a9e47733870889626609d34c5

You can change the zoom direction by changing < and > signs to the other one here: https://gist.github.com/BalazsGyarmati/e199080a9e47733870889626609d34c5#file-scroll_to_plusminus_macos-swift-L172

**Comment by maxrzaw** [comment]
@BalazsGyarmati @alexDevBR Did you use a karabiner command? I set the executable to start at login and it worked great.

**Comment by TylerWhittaker** [comment]
This seems to be the most active issue on adding scroll wheel event support, so I thought it would be worth mentioning here that [the roadmap aka TASKS.md](https://github.com/pqrs-org/Karabiner-Elements/blob/main/TASKS.md) was updated in August 2023 to include "[ ] Add scroll wheel to source events."  I'm optimistic that this means we'll see support for scroll wheel events relatively soon.

**Comment by Noor0** [comment]
Is there a way to map "middle button + scroll" to horizontal scroll? I'm super new to the app and have been searching through the docs to find a way but just can't figure it out.

**Comment by tekezo** [comment]
> middle button + scroll

At the moment, Karabiner-Elements does not support changes to scroll events. While I intend to add support for this eventually, the timeline is yet to be determined.

**Comment by fernando-reis-guimaraes** [comment]
+1

**Comment by BierDav** [comment]
+1

**Comment by mcholbi** [comment]
+1

**Comment by gaabora** [comment]
Bump

**Comment by Jason-K** [comment]
+1
